types/json/binary_functions.go

```diff
@@ -781,6 +781,102 @@ func CompareBinary(left, right BinaryJSON) int {
 	return cmp
 }
 
+// MergePatchBinary implements RFC7396
+// https://datatracker.ietf.org/doc/html/rfc7396
+func MergePatchBinary(bjs []*BinaryJSON) (*BinaryJSON, error) {
+	var err error
+	length := len(bjs)
+
+	// according to the implements of RFC7396
+	// when the last item is not object
+	// we can return the last item directly
+	for i := length - 1; i >= 0; i-- {
+		if bjs[i] == nil || bjs[i].TypeCode != TypeCodeObject {
+			bjs = bjs[i:]
+			break
+		}
+	}
+
+	target := bjs[0]
+	for _, patch := range bjs[1:] {
+		target, err = mergePatchBinary(target, patch)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return target, nil
+}
+
+func mergePatchBinary(target, patch *BinaryJSON) (result *BinaryJSON, err error) {
+	if patch == nil {
+		return nil, nil
+	}
+
+	if patch.TypeCode == TypeCodeObject {
+		if target == nil {
+			return nil, nil
+		}
+
+		keyValMap := make(map[string]BinaryJSON)
+		keys := make([][]byte, 0, len(keyValMap))
+		if target.TypeCode == TypeCodeObject {
+			elemCount := target.GetElemCount()
+			for i := 0; i < elemCount; i++ {
+				key := target.objectGetKey(i)
+				val := target.objectGetVal(i)
+				k := string(key)
+
+				keyValMap[k] = val
+				keys = append(keys, key)
+			}
+		}
+		var tmp *BinaryJSON
+		elemCount := patch.GetElemCount()
+		for i := 0; i < elemCount; i++ {
+			key := patch.objectGetKey(i)
+			val := patch.objectGetVal(i)
+			k := string(key)
+
+			targetKV, exists := keyValMap[k]
+			if val.TypeCode == TypeCodeLiteral && val.Value[0] == LiteralNil {
+				if exists {
+					delete(keyValMap, k)
+				}
+			} else {
+				tmp, err = mergePatchBinary(&targetKV, &val)
+				if err != nil {
+					return result, err
+				}
+
+				if _, ok := keyValMap[k]; !ok {
+					keys = append(keys, key)
+				}
+				keyValMap[k] = *tmp
+			}
+		}
+		sort.Slice(keys, func(i, j int) bool {
+			return bytes.Compare(keys[i], keys[j]) < 0
+		})
+		values := make([]BinaryJSON, 0, len(keys))
+		for index := 0; index < len(keys); {
+			k := string(keys[index])
+			if _, ok := keyValMap[k]; !ok {
+				keys = append(keys[:index], keys[index+1:]...)
+				continue
+			}
```

Comment:
Why we need maintain the slice `keys`?  Can we create the slice using keyValMap after merging?
Step1. Merge two Json by keyValMap
Step2. Create a slice `keys`, iterator all keys in the map and append to the slice.
Step3. Sort keys
Step4. Search values from keyValMap
Step5. buildBinaryObject and return

Reply:
done